qp/ensemble.py

```diff
@@ -0,0 +1,519 @@
+
+    def mix_mod_fit(self, comps=5, using=None, vb=True):
+        """
+        Fits the parameters of a given functional form to an approximation
+
+        Parameters
+        ----------
+        comps: int, optional
+            number of components to consider
+        using: string, optional
+            which existing approximation to use, defaults to first approximation
+        vb: boolean
+            Report progress
+
+        Returns
+        -------
+        self.mix_mod: list, qp.Composite objects
+            list of qp.Composite objects approximating the PDFs
+
+        Notes
+        -----
+        Currently only supports mixture of Gaussians
+        """
+        def mixmod_helper(i):
+            # with open(self.logfilename, 'wb') as logfile:
+            #     logfile.write('fitting pdf '+str(i)+'\n')
+            return self.pdfs[i].mix_mod_fit(n_components=comps, using=using, vb=False)
+
+        self.mix_mod = self.pool.map(mixmod_helper, self.pdf_range)
+
+        return self.mix_mod
+
+    def evaluate(self, loc, using=None, vb=True):
+        """
+        Evaluates all PDFs
+
+        Parameters
+        ----------
+        loc: float or ndarray, float
+            location(s) at which to evaluate the pdfs
+        using: string
+            which parametrization to evaluate, defaults to initialization
+        vb: boolean
+            report on progress
+
+        Returns
+        -------
+        vals: ndarray, ndarray, float
+            the values of the PDFs (or their approximations) at the requested
+            location(s), of shape (npdfs, nlocs)
+        """
+        def evaluate_helper(i):
+            # with open(self.logfilename, 'wb') as logfile:
+            #     logfile.write('evaluating pdf '+str(i)+'\n')
+            return self.pdfs[i].evaluate(loc=loc, using=using, vb=False)
+        self.gridded = self.pool.map(evaluate_helper, self.pdf_range)
+        self.gridded = np.swapaxes(np.array(self.gridded), 0, 1)
+        self.gridded = (self.gridded[0][0], self.gridded[1])
+
+        return self.gridded
+
+    def cdfs(self, limits, using, dx=0.0001):
+        """
+        Computes the integral under the ensemble of PDFs between the given limits.
+        
+        Parameters
+        ----------
+        limits: numpy.ndarray, tuple, float
+            limits of integration, may be different for all PDFs in the ensemble
+        using: string
+            parametrization over which to approximate the integral
+        dx: float, optional
+            granularity of integral
+
+        Returns
+        -------
+        integral: numpy.ndarray, float
+            value of the integral
+        """
+        def integrate_helper(i):
+            return self.pdfs[i].integrate(limits[i], using=using, dx=dx)
+
+        integrals = self.pool.map(integrate_helper, self.pdf_range)
+
+        return integrals
+
+    def stack(self, loc, using, vb=True):
+        """
+        Produces a stack of the PDFs
```

Comment:
It would really be most accurate to call the "stack" an average over PDFs evaluated on a grid.  I added a note explaining this in the docstring and proposing its removal from `qp`'s functionality before an official release, in favor of asking the user to make their own average PDF estimator with `qp.Ensemble.evaluate()`.